Index: kernel-impl/src/main/java/org/sakaiproject/authz/impl/DbAuthzGroupService.java
===================================================================
--- kernel-impl/src/main/java/org/sakaiproject/authz/impl/DbAuthzGroupService.java	(revision 123833)
+++ kernel-impl/src/main/java/org/sakaiproject/authz/impl/DbAuthzGroupService.java	(working copy)
@@ -2443,6 +2443,13 @@
 								active = false;
 							}
 							
+                            for(UserAndRole uar : grants){
+                                if(uar.userId.equals(userId)){
+                                    active = uar.active;
+                                    break;
+                                }
+                            }
+							
 							toInsert.add(new UserAndRole(userId, role, active, true));
 						}
 					}
